File: src/系统/07_任务/任务数据.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local Map = ____lualib.Map
local __TS__New = ____lualib.__TS__New
local __TS__ArrayFrom = ____lualib.__TS__ArrayFrom
local __TS__ArrayFilter = ____lualib.__TS__ArrayFilter
local Set = ____lualib.Set
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local __TS__ArrayEvery = ____lualib.__TS__ArrayEvery
local __TS__ArrayFind = ____lualib.__TS__ArrayFind
local ____exports = {}
--- 获取当前时间戳（War3 Lua 环境下替代 Date.now）
local function now(self)
    return os.time()
end
____exports.QuestType = QuestType or ({})
____exports.QuestType.MAIN = "主线"
____exports.QuestType.SIDE = "支线"
____exports.QuestType.DAILY = "小任务"
--- 任务状态枚举（与War3原生状态对应）
-- 参考：bj_QUESTTYPE_REQ_DISCOVERED, bj_QUESTTYPE_REQ_UNDISCOVERED
____exports.QuestStatus = QuestStatus or ({})
____exports.QuestStatus.UNDISCOVERED = "未发现"
____exports.QuestStatus.DISCOVERED = "已发现"
____exports.QuestStatus.IN_PROGRESS = "进行中"
____exports.QuestStatus.COMPLETED = "已完成"
____exports.QuestStatus.FAILED = "已失败"
--- 任务数据库类
____exports.QuestDatabase = __TS__Class()
local QuestDatabase = ____exports.QuestDatabase
QuestDatabase.name = "QuestDatabase"
function QuestDatabase.prototype.____constructor(self)
    self.quests = __TS__New(Map)
    self.playerData = __TS__New(Map)
end
function QuestDatabase.getInstance(self)
    if not ____exports.QuestDatabase.instance then
        ____exports.QuestDatabase.instance = __TS__New(____exports.QuestDatabase)
    end
    return ____exports.QuestDatabase.instance
end
function QuestDatabase.prototype.registerQuest(self, quest)
    quest.createdAt = quest.createdAt or now(nil)
    quest.updatedAt = quest.updatedAt or now(nil)
    self.quests:set(quest.id, quest)
end
function QuestDatabase.prototype.getQuest(self, id)
    return self.quests:get(id)
end
function QuestDatabase.prototype.getAllQuests(self)
    return __TS__ArrayFrom(self.quests:values())
end
function QuestDatabase.prototype.getQuestsByType(self, ____type)
    return __TS__ArrayFilter(
        __TS__ArrayFrom(self.quests:values()),
        function(____, quest) return quest.type == ____type end
    )
end
function QuestDatabase.prototype.getAvailableQuests(self, playerId, ____type)
    local playerData = self:getPlayerData(playerId)
    if not playerData then
        return {}
    end
    local source = ____type and self:getQuestsByType(____type) or self:getAllQuests()
    return __TS__ArrayFilter(
        source,
        function(____, quest)
            if playerData.quests:has(quest.id) or playerData.completedQuests:has(quest.id) or playerData.failedQuests:has(quest.id) then
                return false
            end
            if quest.requiredQuests and #quest.requiredQuests > 0 then
                for ____, rid in ipairs(quest.requiredQuests) do
                    if not playerData.completedQuests:has(rid) then
                        return false
                    end
                end
            end
            return true
        end
    )
end
function QuestDatabase.prototype.initPlayerData(self, playerId)
    if not self.playerData:has(playerId) then
        self.playerData:set(
            playerId,
            {
                playerId = playerId,
                quests = __TS__New(Map),
                completedQuests = __TS__New(Set),
                failedQuests = __TS__New(Set)
            }
        )
    end
end
function QuestDatabase.prototype.getPlayerData(self, playerId)
    return self.playerData:get(playerId)
end
function QuestDatabase.prototype.acceptQuest(self, playerId, questId)
    local quest = self:getQuest(questId)
    if not quest then
        return false
    end
    local playerData = self:getPlayerData(playerId)
    if not playerData then
        return false
    end
    if playerData.quests:has(questId) or playerData.completedQuests:has(questId) or playerData.failedQuests:has(questId) then
        return false
    end
    if quest.requiredQuests and #quest.requiredQuests > 0 then
        for ____, requiredId in ipairs(quest.requiredQuests) do
            if not playerData.completedQuests:has(requiredId) then
                return false
            end
        end
    end
    local acceptedQuest = __TS__ObjectAssign(
        {},
        quest,
        {
            status = ____exports.QuestStatus.IN_PROGRESS,
            createdAt = now(nil),
            updatedAt = now(nil),
            startTime = now(nil)
        }
    )
    playerData.quests:set(questId, acceptedQuest)
    return true
end
function QuestDatabase.prototype.completeQuest(self, playerId, questId)
    local playerData = self:getPlayerData(playerId)
    if not playerData then
        return false
    end
    local quest = playerData.quests:get(questId)
    if not quest then
        return false
    end
    local allObjectivesCompleted = __TS__ArrayEvery(
        quest.objectives,
        function(____, obj) return obj.completed end
    )
    if not allObjectivesCompleted then
        return false
    end
    quest.status = ____exports.QuestStatus.COMPLETED
    quest.updatedAt = now(nil)
    playerData.quests:delete(questId)
    playerData.completedQuests:add(questId)
    return true
end
function QuestDatabase.prototype.abandonQuest(self, playerId, questId)
    local playerData = self:getPlayerData(playerId)
    if not playerData then
        return false
    end
    local quest = playerData.quests:get(questId)
    if not quest then
        return false
    end
    if quest.status ~= ____exports.QuestStatus.IN_PROGRESS then
        return false
    end
    quest.status = ____exports.QuestStatus.UNDISCOVERED
    playerData.quests:delete(questId)
    return true
end
function QuestDatabase.prototype.failQuest(self, playerId, questId)
    local playerData = self:getPlayerData(playerId)
    if not playerData then
        return false
    end
    local quest = playerData.quests:get(questId)
    if not quest then
        return false
    end
    if quest.status ~= ____exports.QuestStatus.IN_PROGRESS then
        return false
    end
    quest.status = ____exports.QuestStatus.FAILED
    quest.updatedAt = now(nil)
    playerData.quests:delete(questId)
    playerData.failedQuests:add(questId)
    return true
end
function QuestDatabase.prototype.updateObjective(self, playerId, questId, objectiveId, progress)
    local playerData = self:getPlayerData(playerId)
    if not playerData then
        return false
    end
    local quest = playerData.quests:get(questId)
    if not quest then
        return false
    end
    local objective = __TS__ArrayFind(
        quest.objectives,
        function(____, obj) return obj.id == objectiveId end
    )
    if not objective then
        return false
    end
    objective.current = math.min(progress, objective.required)
    objective.completed = objective.current >= objective.required
    quest.updatedAt = now(nil)
    return true
end
function QuestDatabase.prototype.getPlayerActiveQuests(self, playerId)
    local playerData = self:getPlayerData(playerId)
    if not playerData then
        return {}
    end
    return __TS__ArrayFilter(
        __TS__ArrayFrom(playerData.quests:values()),
        function(____, quest) return quest.status == ____exports.QuestStatus.IN_PROGRESS end
    )
end
function QuestDatabase.prototype.getPlayerCompletedQuests(self, playerId)
    local playerData = self:getPlayerData(playerId)
    if not playerData then
        return {}
    end
    return __TS__ArrayFrom(playerData.completedQuests)
end
function QuestDatabase.prototype.getPlayerQuestStatus(self, playerId, questId)
    local playerData = self:getPlayerData(playerId)
    if not playerData then
        return nil
    end
    local quest = playerData.quests:get(questId)
    if quest then
        return quest.status
    end
    if playerData.completedQuests:has(questId) then
        return ____exports.QuestStatus.COMPLETED
    end
    if playerData.failedQuests:has(questId) then
        return ____exports.QuestStatus.FAILED
    end
    return ____exports.QuestStatus.UNDISCOVERED
end
function QuestDatabase.prototype.resetPlayerData(self, playerId)
    self.playerData:delete(playerId)
    self:initPlayerData(playerId)
end
function QuestDatabase.prototype.clearAll(self)
    self.quests:clear()
    self.playerData:clear()
end
____exports.questDB = ____exports.QuestDatabase:getInstance()
function ____exports.createTestQuests(self)
    local db = ____exports.QuestDatabase:getInstance()
    do
        local i = 1
        while i <= 99 do
            local id = "main_" .. (i < 10 and "00" .. tostring(i) or (i < 100 and "0" .. tostring(i) or "" .. tostring(i)))
            local title = "主线任务" .. (i < 10 and "00" .. tostring(i) or (i < 100 and "0" .. tostring(i) or "" .. tostring(i)))
            db:registerQuest({
                id = id,
                type = ____exports.QuestType.MAIN,
                title = title,
                description = "完成基础训练，了解游戏操作",
                objectives = {{
                    id = "obj1",
                    description = "击败训练假人",
                    current = 0,
                    required = 5,
                    completed = false
                }, {
                    id = "obj2",
                    description = "学习技能",
                    current = 0,
                    required = 1,
                    completed = false
                }},
                rewards = {{type = "experience", value = 100, description = "100经验"}, {type = "gold", value = 50, description = "50金币"}},
                status = ____exports.QuestStatus.UNDISCOVERED,
                requiredLevel = 1,
                zone = "新手村",
                icon = i == 2 and "ReplaceableTextures\\CommandButtons\\BTNHeroPaladin.blp" or "ReplaceableTextures\\CommandButtons\\BTNHeroBlademaster.blp",
                createdAt = now(nil),
                updatedAt = now(nil)
            })
            i = i + 1
        end
    end
    db:registerQuest({
        id = "side_002",
        type = ____exports.QuestType.SIDE,
        title = "击杀步兵",
        description = "击杀1个步兵，任务奖励：200金币",
        objectives = {{
            id = "obj1",
            description = "击杀1个步兵",
            current = 0,
            required = 1,
            completed = false
        }},
        rewards = {{type = "gold", value = 200, description = "200金币"}},
        status = ____exports.QuestStatus.UNDISCOVERED,
        requiredLevel = 1,
        zone = "战场",
        icon = "ReplaceableTextures\\CommandButtons\\BTNFootman.blp",
        createdAt = now(nil),
        updatedAt = now(nil)
    })
    db:registerQuest({
        id = "side_001",
        type = ____exports.QuestType.SIDE,
        title = "收集材料",
        description = "为铁匠收集10个铁矿",
        objectives = {{
            id = "obj1",
            description = "收集铁矿",
            current = 0,
            required = 10,
            completed = false
        }},
        rewards = {{type = "item", value = 0, itemId = "item_iron_sword", description = "铁剑"}, {type = "gold", value = 30, description = "30金币"}},
        status = ____exports.QuestStatus.UNDISCOVERED,
        requiredLevel = 3,
        requiredQuests = {"main_001"},
        zone = "矿山",
        icon = "ReplaceableTextures\\CommandButtons\\BTNIronForge.blp",
        createdAt = now(nil),
        updatedAt = now(nil)
    })
    db:registerQuest({
        id = "daily_001",
        type = ____exports.QuestType.DAILY,
        title = "日常巡逻",
        description = "巡逻村庄周边，确保安全",
        objectives = {{
            id = "obj1",
            description = "巡逻指定区域",
            current = 0,
            required = 3,
            completed = false
        }},
        rewards = {{type = "experience", value = 50, description = "50经验"}, {type = "gold", value = 20, description = "20金币"}},
        status = ____exports.QuestStatus.UNDISCOVERED,
        requiredLevel = 2,
        zone = "村庄",
        icon = "ReplaceableTextures\\CommandButtons\\BTNPeon.blp",
        timeLimit = 3600,
        createdAt = now(nil),
        updatedAt = now(nil)
    })
end
return ____exports

File: TS/系统/08．任务系统/任务数据.ts
/**
 * 任务系统 - 数据结构定义和数据管理
 */

/** 获取当前时间戳（War3 Lua 环境下替代 Date.now） */
function now(): number {
  return os.time();
}

export enum QuestType {
  MAIN = "主线",
  SIDE = "支线",
  DAILY = "小任务"
}

/**
 * 任务状态枚举（与War3原生状态对应）
 * 参考：bj_QUESTTYPE_REQ_DISCOVERED, bj_QUESTTYPE_REQ_UNDISCOVERED
 */
export enum QuestStatus {
  UNDISCOVERED = "未发现",    // bj_QUESTTYPE_REQ_DISCOVERED
  DISCOVERED = "已发现",      // bj_QUESTTYPE_REQ_UNDISCOVERED
  IN_PROGRESS = "进行中",
  COMPLETED = "已完成",
  FAILED = "已失败"
}

/**
 * 任务目标接口
 */
export interface QuestObjective {
  id: string;
  description: string;
  current: number;
  required: number;
  completed: boolean;
}

/**
 * 任务奖励接口
 */
export interface QuestReward {
  type: "experience" | "gold" | "item" | "attribute" | "skill";
  value: number;
  itemId?: string;
  description: string;
}

/**
 * 任务数据接口
 */
export interface QuestData {
  id: string;
  type: QuestType;
  title: string;
  description: string;
  objectives: QuestObjective[];
  rewards: QuestReward[];
  status: QuestStatus;
  requiredLevel?: number;
  requiredQuests?: string[];
  zone?: string;
  icon?: string;
  startNpc?: string;
  endNpc?: string;
  timeLimit?: number; // 时间限制（秒），0表示无限制
  createdAt: number; // 任务创建时间戳
  updatedAt: number; // 最后更新时间戳
  nativeHandle?: number; // War3原生任务句柄（CreateQuest返回值）
  startTime?: number; // 任务开始时间戳（用于计时任务）
  /** 已注册但不在任务面板列出（日后启用） */
  uiReserved?: boolean;
}

/**
 * 玩家任务数据接口
 */
export interface PlayerQuestData {
  playerId: number;
  quests: Map<string, QuestData>; // questId -> QuestData
  completedQuests: Set<string>; // 已完成的任务ID
  failedQuests: Set<string>; // 已失败的任务ID
}

/**
 * 任务数据库类
 */
export class QuestDatabase {
  private static instance: QuestDatabase;
  private quests: Map<string, QuestData> = new Map();
  private playerData: Map<number, PlayerQuestData> = new Map();

  private constructor() {}

  public static getInstance(): QuestDatabase {
    if (!QuestDatabase.instance) {
      QuestDatabase.instance = new QuestDatabase();
    }
    return QuestDatabase.instance;
  }

  /**
   * 注册任务定义
   */
  public registerQuest(quest: QuestData): void {
    quest.createdAt = quest.createdAt || now();
    quest.updatedAt = quest.updatedAt || now();
    this.quests.set(quest.id, quest);
  }

  /**
   * 获取任务定义
   */
  public getQuest(id: string): QuestData | undefined {
    return this.quests.get(id);
  }

  /**
   * 获取所有任务定义
   */
  public getAllQuests(): QuestData[] {
    return Array.from(this.quests.values());
  }

  /**
   * 根据类型获取任务定义
   */
  public getQuestsByType(type: QuestType): QuestData[] {
    return Array.from(this.quests.values()).filter(quest => quest.type === type);
  }

  /**
   * 获取玩家可接任务（未接受、前置已完成）
   */
  public getAvailableQuests(playerId: number, type?: QuestType): QuestData[] {
    const playerData = this.getPlayerData(playerId);
    if (!playerData) return [];

    const source = type ? this.getQuestsByType(type) : this.getAllQuests();
    return source.filter(quest => {
      if (quest.uiReserved) return false;
      if (playerData.quests.has(quest.id) ||
          playerData.completedQuests.has(quest.id) ||
          playerData.failedQuests.has(quest.id)) {
        return false;
      }
      if (quest.requiredQuests && quest.requiredQuests.length > 0) {
        for (const rid of quest.requiredQuests) {
          if (!playerData.completedQuests.has(rid)) return false;
        }
      }
      return true;
    });
  }

  /**
   * 初始化玩家任务数据
   */
  public initPlayerData(playerId: number): void {
    if (!this.playerData.has(playerId)) {
      this.playerData.set(playerId, {
        playerId,
        quests: new Map(),
        completedQuests: new Set(),
        failedQuests: new Set()
      });
    }
  }

  /**
   * 获取玩家任务数据
   */
  public getPlayerData(playerId: number): PlayerQuestData | undefined {
    return this.playerData.get(playerId);
  }

  /**
   * 玩家接受任务
   */
  public acceptQuest(playerId: number, questId: string): boolean {
    const quest = this.getQuest(questId);
    if (!quest || quest.uiReserved) return false;

    const playerData = this.getPlayerData(playerId);
    if (!playerData) return false;

    // 检查是否已经接受或完成
    if (playerData.quests.has(questId) ||
        playerData.completedQuests.has(questId) ||
        playerData.failedQuests.has(questId)) {
      return false;
    }

    // 检查前置任务
    if (quest.requiredQuests && quest.requiredQuests.length > 0) {
      for (const requiredId of quest.requiredQuests) {
        if (!playerData.completedQuests.has(requiredId)) {
          return false;
        }
      }
    }

    // 克隆任务数据并设置为进行中
    const acceptedQuest: QuestData = {
      ...quest,
      status: QuestStatus.IN_PROGRESS,
      createdAt: now(),
      updatedAt: now(),
      startTime: now(),
    };

    playerData.quests.set(questId, acceptedQuest);
    return true;
  }

  /**
   * 玩家完成任务
   */
  public completeQuest(playerId: number, questId: string): boolean {
    const playerData = this.getPlayerData(playerId);
    if (!playerData) return false;

    const quest = playerData.quests.get(questId);
    if (!quest) return false;

    // 检查所有目标是否完成
    const allObjectivesCompleted = quest.objectives.every(obj => obj.completed);
    if (!allObjectivesCompleted) return false;

    // 更新状态
    quest.status = QuestStatus.COMPLETED;
    quest.updatedAt = now();

    // 移动到已完成列表
    playerData.quests.delete(questId);
    playerData.completedQuests.add(questId);

    return true;
  }

  /**
   * 玩家放弃任务
   */
  public abandonQuest(playerId: number, questId: string): boolean {
    const playerData = this.getPlayerData(playerId);
    if (!playerData) return false;

    const quest = playerData.quests.get(questId);
    if (!quest) return false;
    if (quest.status !== QuestStatus.IN_PROGRESS) return false;

    quest.status = QuestStatus.UNDISCOVERED;
    playerData.quests.delete(questId);
    return true;
  }

  /**
   * 玩家任务失败
   */
  public failQuest(playerId: number, questId: string): boolean {
    const playerData = this.getPlayerData(playerId);
    if (!playerData) return false;

    const quest = playerData.quests.get(questId);
    if (!quest) return false;
    if (quest.status !== QuestStatus.IN_PROGRESS) return false;

    quest.status = QuestStatus.FAILED;
    quest.updatedAt = now();
    playerData.quests.delete(questId);
    playerData.failedQuests.add(questId);
    return true;
  }

  /**
   * 更新任务目标进度
   */
  public updateObjective(playerId: number, questId: string, objectiveId: string, progress: number): boolean {
    const playerData = this.getPlayerData(playerId);
    if (!playerData) return false;

    const quest = playerData.quests.get(questId);
    if (!quest) return false;

    const objective = quest.objectives.find(obj => obj.id === objectiveId);
    if (!objective) return false;

    objective.current = Math.min(progress, objective.required);
    objective.completed = objective.current >= objective.required;
    quest.updatedAt = now();

    return true;
  }

  /**
   * 获取玩家进行中的任务
   */
  public getPlayerActiveQuests(playerId: number): QuestData[] {
    const playerData = this.getPlayerData(playerId);
    if (!playerData) return [];

    return Array.from(playerData.quests.values()).filter(
      quest => quest.status === QuestStatus.IN_PROGRESS
    );
  }

  /**
   * 获取玩家已完成的任务
   */
  public getPlayerCompletedQuests(playerId: number): string[] {
    const playerData = this.getPlayerData(playerId);
    if (!playerData) return [];

    return Array.from(playerData.completedQuests);
  }

  /**
   * 获取玩家任务状态
   */
  public getPlayerQuestStatus(playerId: number, questId: string): QuestStatus | undefined {
    const playerData = this.getPlayerData(playerId);
    if (!playerData) return undefined;

    const quest = playerData.quests.get(questId);
    if (quest) return quest.status;

    if (playerData.completedQuests.has(questId)) return QuestStatus.COMPLETED;
    if (playerData.failedQuests.has(questId)) return QuestStatus.FAILED;

    return QuestStatus.UNDISCOVERED;
  }

  /**
   * 重置玩家任务数据（用于测试）
   */
  public resetPlayerData(playerId: number): void {
    this.playerData.delete(playerId);
    this.initPlayerData(playerId);
  }

  /**
   * 清除所有数据（用于测试）
   */
  public clearAll(): void {
    this.quests.clear();
    this.playerData.clear();
  }
}

// 导出单例实例
export const questDB = QuestDatabase.getInstance();

// 测试数据（开发阶段使用）
export function createTestQuests(): void {
  const db = QuestDatabase.getInstance();

  for (let i = 1; i <= 99; i++) {
    const id = "main_" + (i < 10 ? "00" + i : i < 100 ? "0" + i : "" + i);
    const title = "主线任务" + (i < 10 ? "00" + i : i < 100 ? "0" + i : "" + i);
    db.registerQuest({
      id,
      type: QuestType.MAIN,
      title,
      description: "完成基础训练，了解游戏操作",
      objectives: [
        { id: "obj1", description: "击败训练假人", current: 0, required: 5, completed: false },
        { id: "obj2", description: "学习技能", current: 0, required: 1, completed: false }
      ],
      rewards: [
        { type: "experience", value: 100, description: "100经验" },
        { type: "gold", value: 50, description: "50金币" }
      ],
      status: QuestStatus.UNDISCOVERED,
      requiredLevel: 1,
      zone: "新手村",
      icon:
        i === 2
          ? "ReplaceableTextures\\CommandButtons\\BTNHeroPaladin.blp"
          : "ReplaceableTextures\\CommandButtons\\BTNHeroBlademaster.blp",
      createdAt: now(),
      updatedAt: now()
    });
  }

  // 支线任务：击杀步兵
  db.registerQuest({
    id: "side_002",
    type: QuestType.SIDE,
    title: "击杀步兵",
    description: "击杀1个步兵，任务奖励：200金币",
    objectives: [
      { id: "obj1", description: "击杀1个步兵", current: 0, required: 1, completed: false }
    ],
    rewards: [
      { type: "gold", value: 200, description: "200金币" }
    ],
    status: QuestStatus.UNDISCOVERED,
    requiredLevel: 1,
    zone: "战场",
    icon: "ReplaceableTextures\\CommandButtons\\BTNFootman.blp",
    createdAt: now(),
    updatedAt: now()
  });

  // 支线任务示例（保留）
  db.registerQuest({
    id: "side_001",
    type: QuestType.SIDE,
    title: "收集材料",
    description: "为铁匠收集10个铁矿",
    objectives: [
      { id: "obj1", description: "收集铁矿", current: 0, required: 10, completed: false }
    ],
    rewards: [
      { type: "item", value: 0, itemId: "item_iron_sword", description: "铁剑" },
      { type: "gold", value: 30, description: "30金币" }
    ],
    status: QuestStatus.UNDISCOVERED,
    requiredLevel: 3,
    requiredQuests: ["main_001"],
    zone: "矿山",
    icon: "ReplaceableTextures\\CommandButtons\\BTNIronForge.blp",
    createdAt: now(),
    updatedAt: now()
  });

  // 占位：支线 / 小任务各一条，结构与主线列表行一致（`side_`/`daily_` + 001–020 带图标）；日后去掉 `uiReserved` 即可启用
  db.registerQuest({
    id: "side_003",
    type: QuestType.SIDE,
    title: "（占位）支线任务",
    description: "日后启用",
    objectives: [
      { id: "obj1", description: "占位目标", current: 0, required: 1, completed: false }
    ],
    rewards: [{ type: "gold", value: 0, description: "" }],
    status: QuestStatus.UNDISCOVERED,
    requiredLevel: 1,
    zone: "",
    icon: "ReplaceableTextures\\CommandButtons\\BTNFootman.blp",
    uiReserved: true,
    createdAt: now(),
    updatedAt: now()
  });

  db.registerQuest({
    id: "daily_002",
    type: QuestType.DAILY,
    title: "（占位）小任务",
    description: "日后启用",
    objectives: [
      { id: "obj1", description: "占位目标", current: 0, required: 1, completed: false }
    ],
    rewards: [{ type: "gold", value: 0, description: "" }],
    status: QuestStatus.UNDISCOVERED,
    requiredLevel: 1,
    zone: "",
    icon: "ReplaceableTextures\\CommandButtons\\BTNPeon.blp",
    uiReserved: true,
    createdAt: now(),
    updatedAt: now()
  });

  // 小任务示例
  db.registerQuest({
    id: "daily_001",
    type: QuestType.DAILY,
    title: "日常巡逻",
    description: "巡逻村庄周边，确保安全",
    objectives: [
      { id: "obj1", description: "巡逻指定区域", current: 0, required: 3, completed: false }
    ],
    rewards: [
      { type: "experience", value: 50, description: "50经验" },
      { type: "gold", value: 20, description: "20金币" }
    ],
    status: QuestStatus.UNDISCOVERED,
    requiredLevel: 2,
    zone: "村庄",
    icon: "ReplaceableTextures\\CommandButtons\\BTNPeon.blp",
    timeLimit: 3600, // 1小时
    createdAt: now(),
    updatedAt: now()
  });
}